### ISSUE TITLE
Add error for incomplete enumeration specification

### ISSUE THREAD

**Issue by katherine-h** [uber/pyro#946]
To enumerate over discrete variables, it's necessary both to supply the argument `enum_discrete=True` to `SVI`, and to mark discrete samples using `infer = {"enumerate": "sequential"}` (or, equivalently, wrap the guide in `config_enumerate`). However, I noticed that the model runs (albeit without enumeration) if one marks `enum_discrete=True` but neglects to mark the samples. I believe this should throw an error: with the current setup, it's easy to think you're running a model with discrete enumeration when you're actually not.

**Comment by eb8680** [comment]
The current behavior is by design, since there are cases where we'd like to be able to enumerate some variables and sample others, even if they could be enumerated.  It might be a good idea add a warning to go along with the default behavior, though.

**Comment by neerajprad** [comment]
+1. I think @rohitsingh0812 was also confused about this behavior. We should include a warning specially with `pyro.set_validation(True)`.

**Comment by katherine-h** [comment]
@eb8680 Yes, I understand, but it seems like it should check that there's at least one instance of a sample marked for enumeration.

**Comment by fritzo** [comment]
I think it would be helpful to warn in all potentially mismatched cases:
- when `enum_discrete=True` but no sites are marked for enumeration
- when `enum_discrete=False` but some sites are marked for enumeration

Also @eb8680 suggested revising the `SVI` interface and @jpchen suggested renaming `config_enumerate` to `mark_sites_for_enumeration`, both of which should avoid confusion.
